Makefile

```diff
@@ -1,261 +1,91 @@
-IMPORTER_BUILD=$(BUILD_DIR)/$(IMPORTER)
-
-# DOCKER TAG VARS
-DEV_REGISTRY=jcoperh
-RELEASE_REGISTRY=kubevirt
-RELEASE_TAG=$(shell git describe --tags --abbrev=0 HEAD)
-CTRL_IMG_NAME=cdi-$(CONTROLLER)
-IMPT_IMG_NAME=cdi-$(IMPORTER)
-GIT_USER=$(shell git config --get user.email | sed 's/@.*//')
-TAG=$(GIT_USER)-latest
-
-# Preflight Check Defaults
-USE_DOCKER=1
-RUNNING_DOCKER=$(shell docker ps > /dev/null 2>&1; echo $$?)
-
-.PHONY: controller importer controller-bin importer-bin controller-image importer-image push-controller push-controller-release push-importer-release push-importer lib clean test
-all: clean test controller importer lib
-controller: controller-bin controller-image
-importer: importer-bin importer-image
-push: push-importer push-controller
-test: functional-test unit-test
-test-local: unit-test-local
-functional-test: func-test-bin func-test-image func-test-run
-unit-test: unit-test-bin unit-test-image-controller unit-test-image-image unit-test-run
-lib: lib-size
-
-BUILD_IMAGE=golang:1.10.2
-WORK_DIR=/go/src/kubevirt.io/containerized-data-importer
-GOOS?=linux
-ARCH?=amd64
-CGO_ENABLED=0
-LDFLAGS='-extldflags "-static"'
-
-# Compile controller binary
-controller-bin:
-	@echo '********'
-	@echo 'Compiling controller binary'
-	@if [ '$(RUNNING_DOCKER)' -eq 0 ] && [ $(USE_DOCKER) -eq 1 ]; then \
-		echo 'building with docker'; \
-		$(DOCKER_BUILD_CMD) -o $(WORK_DIR)/bin/$(CONTROLLER_BIN) $(WORK_DIR)/cmd/controller/controller.go; \
-	else \
-		echo 'building without docker'; \
-		$(BUILD_CMD) -o $(REPO_ROOT)/bin/$(CONTROLLER_BIN) $(CONTROLLER_CMD)/controller.go; \
-	fi
-
-# Compile importer binary
-importer-bin:
-	@echo '********'
-	@echo 'Compiling importer binary'
-	@if [ $(RUNNING_DOCKER) -eq 0 ] && [ $(USE_DOCKER) -eq 1 ]; then \
-		echo 'building with docker'; \
-		$(DOCKER_BUILD_CMD) -o $(WORK_DIR)/bin/$(IMPORTER_BIN) $(WORK_DIR)/cmd/importer/importer.go; \
-	else \
-		echo 'building without docker'; \
-		$(BUILD_CMD) -o $(REPO_ROOT)/bin/$(IMPORTER_BIN) $(IMPORTER_CMD)/importer.go; \
-	fi
-
-# Compile datastream functional test binary
-func-test-bin:
-	@echo '********'
-	@echo 'Compiling functional test binary'
-	-rm -f $(F_TEST_BIN)
-	GOOS=$(GOOS) GOARCH=$(ARCH) CGO_ENABLED=$(CGO_ENABLED) go test -a -c -ldflags $(LDFLAGS) -o $(F_TEST_BIN) $(F_TEST_DIR)/*.go
-
-# Compile datastream functional test binary
-unit-test-bin:
-	@echo '********'
-	@echo 'Compiling unit test binary'
-	-rm -f $(U_TEST_BIN)
-	-rm -f $(U_TEST_BIN_CONTROLLER)
-	-rm -f $(U_TEST_BIN_IMAGE)
-	GOOS=$(GOOS) GOARCH=$(ARCH) CGO_ENABLED=$(CGO_ENABLED) go test -c -v -tags=unit_test ./pkg/controller -a -ldflags $(LDFLAGS) -o $(U_TEST_BIN_CONTROLLER) $(U_TEST_DIR_ALL)/*_test.go
-	GOOS=$(GOOS) GOARCH=$(ARCH) CGO_ENABLED=$(CGO_ENABLED) go test -c -v -tags=unit_test ./pkg/image -a -ldflags $(LDFLAGS) -o $(U_TEST_BIN_IMAGE) $(U_TEST_DIR_ALL)/*_test.go
+clean:
+	./hack/build/in-docker "./hack/build/build-go.sh clean && rm -rf bin/* _out/*"
```

Comment:
We always clean in docker even if DOCKER=0?  Will that work?

Reply:
The reason is that when built with a container, the files will have `root:root` ownership, since the container runs as root internally.  This means that if make clean is executed in the host env, not in a container, it won't have privilege to delete the files.  (ownership translation is apparently something people have been asking docker for for years)

That said, I'm not against adding the non-dockerized version.  It will succeed the build is only ever performed with DOCKER=0.